### PR TITLE
planner, executor: reset NotNullFlag when merge schema for join

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/22941 

Problem Summary:
Root cause: The not null flag of a column is set to true wrongly even if this column is from the inner child of an outer join. Suppose the column is `a`, when calling `foldConstant(isnull(a))`, it'll be optimized to `0` because the optimizer thinks `a` will never be null.

### What is changed and how it works?

What's Changed:
Reset notNullFlag in mergeSchema.

How it Works:

1. Before this commit, [buildJoin](https://github.com/pingcap/tidb/blob/master/planner/core/logical_plan_builder.go#L710) will `resetNotNullFlag` for outer join. 
2. When doing column pruning, we'll [rebuild the schema](https://github.com/pingcap/tidb/blob/master/planner/core/rule_column_pruning.go#L319) for join using the schema of its children directly, thus the `notNullFlag` reset operation before is **rollbacked**.
3. In normal cases, `isnull` function is built when building the logical plan. Thus this bug will not be triggered because we'll not call foldConstant on isnull function.
4. But when the `projection elimination` rule is called, [foldConstant](https://github.com/pingcap/tidb/blob/master/planner/core/rule_eliminate_projection.go#L177) will be invoked on the `isnull` funtion, and the isnull is optimized to 0 wrongly.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

N/A

### Release note 

- Fix the bug when the argument of the `is null` function is the inner side of an outer join.

### CODE REVIEW

**Review by lzmhhh123 — APPROVED** [review]
/LGTM

**Review by rebelice — APPROVED** [review]

**Comment by fzhedu** [comment]
still a problem:
reset the not null flag too earily, but the outer join may be optimized into inner join, so some not-null values are set nullable wrongly.

File: planner/core/rule_column_pruning.go [hunk 1]
```diff
@@ -317,6 +317,13 @@ func (p *LogicalJoin) mergeSchema() {
 		p.schema.Append(joinCol)
 	} else {
 		p.schema = expression.MergeSchema(lChild.Schema(), rChild.Schema())
+		switch p.JoinType {
+		case LeftOuterJoin:
+			resetNotNullFlag(p.schema, p.children[1].Schema().Len(), p.schema.Len())
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply planner/core/rule_column_pruning.go:322]: ref https://github.com/pingcap/tidb/blob/505955ad0b6bdaf2550c1bd1ab4858834af02048/planner/core/util.go#L228

**Pull request merged**